src/rgw/driver/d4n/d4n_policy.cc

```diff
@@ -0,0 +1,411 @@
+    request req;
+    req.push("HSET", "lfuda", "age", std::to_string(age));
+
+    redis_exec(conn, ec, req, resp, y);
+
+    if (ec)
+      return {};
+
+    return std::get<0>(resp).value(); /* Returns number of fields set */
+  } catch(std::exception &e) {
+    return -1;
+  }
+}
+
+int LFUDAPolicy::get_age(optional_yield y) {
+  response<int> resp;
+
+  try {
+    boost::system::error_code ec;
+    request req;
+    req.push("HEXISTS", "lfuda", "age");
+
+    redis_exec(conn, ec, req, resp, y);
+
+    if (ec)
+      return -1;
+  } catch(std::exception &e) {
+    return -1;
+  }
+
+  if (!std::get<0>(resp).value()) {
+    if (set_age(0, y)) /* Initialize age */
+      return 0;
+    else
+      return -1;
+  }
+
+  try { 
+    boost::system::error_code ec;
+    response<std::string> value;
+    request req;
+    req.push("HGET", "lfuda", "age");
+      
+    redis_exec(conn, ec, req, value, y);
+
+    if (ec)
+      return -1;
+
+    return std::stoi(std::get<0>(value).value());
+  } catch(std::exception &e) {
+    return -1;
+  }
+}
+
+int LFUDAPolicy::set_min_avg_weight(size_t weight, std::string cacheLocation, optional_yield y) {
+  try {
+    boost::system::error_code ec;
+    response<int> resp;
+    request req;
+    req.push("HSET", "lfuda", "minAvgWeight:cache", cacheLocation);
+
+    redis_exec(conn, ec, req, resp, y);
+
+    if (ec)
+      return {};
+  } catch(std::exception &e) {
+    return -1;
+  }
+  
+  try {
+    boost::system::error_code ec;
+    response<int> resp;
+    request req;
+    req.push("HSET", "lfuda", "minAvgWeight:weight", boost::lexical_cast<int>(weight));
+
+    redis_exec(conn, ec, req, resp, y);
+
+    if (ec)
+      return {};
+
+    return std::get<0>(resp).value(); /* Returns number of fields set */
+  } catch(std::exception &e) {
+    return -1;
+  }
+}
+
+int LFUDAPolicy::get_min_avg_weight(optional_yield y) {
+  response<int> resp;
+
+  try {
+    boost::system::error_code ec;
+    request req;
+    req.push("HEXISTS", "lfuda", "minAvgWeight:cache");
+
+    redis_exec(conn, ec, req, resp, y);
+
+    if (ec)
+      return -1;
+  } catch(std::exception &e) {
+    return -1;
+  }
+
+  if (!std::get<0>(resp).value()) {
+    if (set_min_avg_weight(0, dir->cct->_conf->rgw_local_cache_address, y)) { /* Initialize minimum average weight */
+      return 0;
+    } else {
+      return -1;
+    }
+  }
+
+  try { 
+    boost::system::error_code ec;
+    response<std::string> value;
+    request req;
+    req.push("HGET", "lfuda", "minAvgWeight:weight");
+      
+    redis_exec(conn, ec, req, value, y);
+
+    if (ec)
+      return -1;
+
+    return std::stoi(std::get<0>(value).value());
+  } catch(std::exception &e) {
+    return -1;
+  }
+}
+
+CacheBlock LFUDAPolicy::find_victim(const DoutPrefixProvider* dpp, rgw::cache::CacheDriver* cacheNode, optional_yield y) {
+  if (entries_map.empty())
+    return {};
+
+  auto it = std::min_element(std::begin(entries_map), std::end(entries_map),
```

Comment:
What if there are two blocks with the same localWeight?

Reply:
O(n) traversal of the map isn't very efficient in case the map is huge, in that case a data structure ordered by localWeight could be useful, so that the victim block can be looked up in O(1) time.